Perform reverse NAT at Host Interface

krishgobinath opened the pull request:
Description: if NETFILTER_COMPAT_MODE flag is set, guarantees that traffic will
pass-through kernel netfilter and that Iptables rules are enforced on traffic.

Problem: if Nodeport traffic ingress on the same node as a backend pod's node,
NAT and reverse NAT translations at 2 different interfaces.
On ingress path, NAT happens at Node's host interface, pass through kernel stack.
on egress path reverse NAT happen at Pod's Veth interface and bypasses kernel stack.
This asymmetric behavior, reflects in kernel conntrack entries and having a IPTABLE
rule to DROP any packets if conntrack state is INVALID which results in packet DROP.

Fixes: #11914

Signed-off-by: Gobinath Krishnamoorthy <gobinathk@google.com>

This PR was already reviewed and merged [here](https://github.com/cilium/cilium/pull/15354), we found some flakiness in CI tests, which was added by this [PR](https://github.com/cilium/cilium/pull/15354), hence this PR was reverted. Github Issue for flakiness tracked [here](https://github.com/cilium/cilium/issues/17060). 

Added fix In test/k8sT/Services.go and test/helpers/kubectl.go to resolve the flakiness issue.
@brb @pchaigno

Commit:
Perform reverse NAT at Host Interface

Description: if NETFILTER_COMPAT_MODE flag is set, guarantees that traffic will
pass-through kernel netfilter and that Iptables rules are enforced on traffic.

Problem: if Nodeport traffic ingress on the same node as a backend pod's node,
NAT and reverse NAT translations at 2 different interfaces.
On ingress path, NAT happens at Node's host interface, pass through kernel stack.
on egress path reverse NAT happen at Pod's Veth interface and bypasses kernel stack.
This asymmetric behavior, reflects in kernel conntrack entries and having a IPTABLE
rule to DROP any packets if conntrack state is INVALID which results in packet DROP.

Fixes: #11914

Signed-off-by: Gobinath Krishnamoorthy <gobinathk@google.com>
Signed-off-by: Weilong Cui <cuiwl@google.com>

pchaigno commented on daemon/cmd/daemon_main.go:
```diff
@@ -1452,6 +1456,15 @@ func initEnv(cmd *cobra.Command) {
+	// NetfilterCompatibleMode should be enabled only on higher versions (5.4 or later) of kernel.
+	// If kernel doesn't support higher instruction complexity limit, then disable NetfilterCompatibleMode.
+	if option.Config.NetfilterCompatibleMode {
+		if !probes.NewProbeManager().GetMisc().HaveLargeInsnLimit {
```
This actually corresponds to [v5.2](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=c04c0d2b968ac45d6ef020316808ef6c82325a82), so we should update the comment and warning message.

krishgobinath replied:
Done.

pchaigno commented on pkg/datapath/linux/config/config.go:
```diff
@@ -390,6 +391,17 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
+	// this feature only in kernels support higher instruction complexity limit.
+	// NetfilterCompatibleMode feature can be enabled on all kernels, once we have support
+	// for adding new tail calls in host device[bpf_host].
+	if option.Config.NetfilterCompatibleMode && probes.NewProbeManager().GetMisc().HaveLargeInsnLimit &&
```
We already checked kernel support during the agent bootstrap in `daemon/cmd/daemon_main.go`, so why do we need to check again here?

krishgobinath replied:
I don't recall why this check was added, looks like redundant. Updated PR.

pchaigno reviewed (changes requested):
LGTM apart from a couple minor comments.

I think the fix for the tests would however merit its own commit, with a proper explanation of what's happening in the commit description.

christarazi commented on test/k8sT/Services.go:
```diff
@@ -636,6 +636,35 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+				})
+
+				AfterAll(func() {
+					DeployCiliumAndDNS(kubectl, ciliumFilename)
```
Why is Cilium being redeployed after the test runs?

pchaigno replied:
@christarazi To restore the default configuration and ensure subsequent tests are not executed with `netfilterCompatMode=true` I'm guessing.

christarazi replied:
Makes sense, not every test will reconfigure Cilium

krishgobinath replied:
The main reason to revert `netfilterCompatMode` configuration is to avoid issues like mentioned in this [GH issue](https://github.com/cilium/cilium/issues/17060). Before actual endpoint regeneration completes, we start running the test(s). If the current `bpf host` and previous installation of `bpf host` differs significantly we might run into some issues.

christarazi commented on test/helpers/kubectl.go:
```diff
@@ -4485,3 +4485,36 @@ func (kub *Kubectl) NslookupInPod(namespace, pod string, target string) (err err
+
+	return WithTimeout(
+		body,
+		fmt.Sprintf("timed out waiting for cilium pod :%s to regenerate", ciliumPod),
```
```suggestion
		fmt.Sprintf("timed out waiting for cilium pod %q to regenerate", ciliumPod),
```

christarazi commented on test/helpers/kubectl.go:
```diff
@@ -4485,3 +4485,36 @@ func (kub *Kubectl) NslookupInPod(namespace, pod string, target string) (err err
+
+	return WithTimeout(
+		body,
+		fmt.Sprintf("timed out waiting for cilium pod :%q to regenerate", ciliumPod),
```
Nit: the `:` is superfluous 

```suggestion
		fmt.Sprintf("timed out waiting for cilium pod %q to regenerate", ciliumPod),
```

krishgobinath replied:
Done.

christarazi reviewed (approved):
LGTM, one minor nit

pchaigno reviewed (changes requested):
This still needs to be split in two commits per https://github.com/cilium/cilium/pull/17504#pullrequestreview-767622519.

Commit:
Add E2E tests for validating netfilter-compat-mode feature

Added few e2e tests to validate netfilter-compatible feature, these
tests are set to run only if cilium endpoint regeneration completes
successfully. Implemented WaitForCiliumPodToCompleteRegeneration  helper function,
waits and ensures endpoint regeneration completes successfully.

Signed-off-by: Gobinath Krishnamoorthy <gobinathk@google.com>

kaworu commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -749,6 +749,10 @@ data:
 {{- end }}
 
+{{- if hasKey .Values "netfilterCompatMode" }}
+  netfilter-compatible-mode: {{ .Values.netfilterCompatMode | quote }}
```
Please add `netfilterCompatMode` to `install/kubernetes/cilium/values.yaml` with a top comment starting with `--` (see the other values). Then run  `make -C install/kubernetes docs` in order to update `install/kubernetes/cilium/README.md`.

krishgobinath replied:
Done.

kkourt commented on test/helpers/utils.go:
```diff
@@ -527,6 +527,18 @@ func DoesNotRunOn419OrLaterKernel() bool {
+// 5.4 or net-next kernels.
+func RunsOn54OrLaterKernel() bool {
+	return RunsOnNetNextKernel() || RunsOn54Kernel()
+}
```
Do these functions belong in the second patch?

krishgobinath replied:
Looks like it was part of another commit.Fixed correctly now.

kkourt reviewed (changes requested):
Hi, 

Thanks for the patch.

One thing that it is not clear to me is the following: `NETFILTER_COMPAT_MODE` seems to imply that _all_ traffic would pass via netfilter. I understand that the  above flag was not in the original patch, but was suggested by @brb in https://github.com/cilium/cilium/pull/15354#issuecomment-814738144. Here's my question: is this patch supposed to ensure that _all_ traffic will go over netfilter, or only nodeport traffic? Because at least in `bpf_lxc.c` all the changes seem to happen under a `if (ct_state.node_port)` guard.

Also, is the inclusion of the binary files `test/bpf/drop-notify-test` and `test/bpf/nat-test` intentional?

pchaigno commented on bpf/Makefile:
```diff
@@ -172,6 +172,11 @@ MAX_HOST_OPTIONS = $(MAX_BASE_OPTIONS) -DENCAP_IFINDEX=1 -DTUNNEL_MODE=1
 ifneq ("$(KERNEL)","49")
 MAX_HOST_OPTIONS += -DENABLE_EGRESS_GATEWAY=1
 endif
+ifeq ("$(KERNEL)","54")
+MAX_HOST_OPTIONS += -DNETFILTER_COMPAT_MODE=1
+else ifeq ("$(KERNEL)","netnext")
+MAX_HOST_OPTIONS += -DNETFILTER_COMPAT_MODE=1
+endif
```
```suggestion
ifneq ("$(KERNEL)","419")
MAX_HOST_OPTIONS += -DNETFILTER_COMPAT_MODE=1
endif
```

krishgobinath replied:
We don't want this feature for kernel version "49"  as well.

pchaigno replied:
Ah, right! I thought we were in the else branch of 49. Parsing Makefiles is hard... :-)

pchaigno commented on test/helpers/kubectl.go:
```diff
@@ -4485,3 +4485,36 @@ func (kub *Kubectl) NslookupInPod(namespace, pod string, target string) (err err
+// WaitForCiliumPodToCompleteRegeneration waits for the cilium endpoint regneration to succeed, until timeout occurs.
+// Endpoint.status.log tracks the regeneration status, if log.Code equals OK and state equals ready
+// then it is considered as success.
+func (kub *Kubectl) WaitForCiliumPodToCompleteRegeneration(ciliumPod string, timeout time.Duration) error {
```
Could you explain in the comment or in the commit description why this is now needed and wasn't needed for any other test before?

krishgobinath replied:
Updated comment for `WaitForCiliumPodToCompleteRegeneration`

pchaigno reviewed (changes requested):
It looks like you committed two files by mistake: `test/bpf/drop-notify-test` and `test/bpf/nat-test`.

pchaigno commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -306,6 +312,11 @@ func (p *ProbeManager) GetMapTypes() *MapTypes {
+// GetMisc returns information about miscellaneous eBPF features.
+func (p *ProbeManager) GetMisc() *Misc {
+	return &p.features.Misc
+}
```
Another contributor added this probe because they also needed it. That's why you now have a merge conflict.

krishgobinath replied:
resolved.

pchaigno commented on test/helpers/kubectl.go:
```diff
@@ -4485,3 +4485,42 @@ func (kub *Kubectl) NslookupInPod(namespace, pod string, target string) (err err
+// without NetfilterCompatibleMode enabled, the reverse DNAT for reply packet is done at bpf_lxc. If we don't wait for
+// endpoint regneration to complete, there may be race condition cases, where the reverse DNAT for reply packet is
+// may not performed by both bpf_lxc and bpf_host.
+// Hence, we need to wait for endpoint regeneration to complete on endpoints.
```
I think these six lines should be moved to the call site, in `test/k8sT/Services.go`.

krishgobinath replied:
Comment updated.

pchaigno commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -312,9 +312,9 @@ func (p *ProbeManager) GetMapTypes() *MapTypes {
-	return p.features.Misc
+// GetMisc returns information about miscellaneous eBPF features.
+func (p *ProbeManager) GetMisc() *Misc {
+	return &p.features.Misc
```
Why this change?

krishgobinath replied:
The caller of this function reads only fields in `Misc`.  Returning pointer to the object is suffice here.
https://github.com/cilium/cilium/pull/17504/files#diff-c0e8dc92b48dcab08b42d07a5223eb03c267c8484f975a66f423ae85d9cc6ea2R1505

pchaigno replied:
See https://github.com/cilium/cilium/pull/17541#discussion_r728683277.

krishgobinath replied:
I'm curious, I'm not compiler expert and wondering how returning a pointer will allocate from heap? Here is what i see equivalent assembly code for this function

```
func (p *ProbeManager) GetMisc() *Misc {
	return &p.features.Misc
	}
```
```
6137908 TEXT github.com/cilium/cilium/pkg/datapath/linux/probes.(*ProbeManager).GetMisc(SB) /usr/local/google/home/gobinathk/revnat-fix/cilium/pkg/datapath/linux/probes/probes.go
6137909         return &p.features.Misc
6137910   0x1a927a0             8400                    TESTB AL, 0(AX)
6137911   0x1a927a2             480558020000            ADDQ $0x258, AX
6137912   0x1a927a8             c3                      RET
```

joamaki replied:
The heap allocation here depends not only on what happens to `*Misc`, but also on where `ProbeManager` is instantiated and used. Here it seems like `p` already escapes the scope in which it is defined and hence changing `Misc` to `*Misc` does not cause heap allocation and we also don't need to copy its contents.

However, since `Misc` is a tiny struct and we certainly don't want the caller of `GetMisc` to modify its contents, I would recommend keeping it as it was and not return a pointer.

pchaigno replied:
cc @ti-mo who initially suggested this change.

ti-mo replied:
Ah, of course, the `Misc` wasn't allocated in this function and already escaped from the start. Regardless, we don't want to return a pointer here.

pchaigno commented on test/k8sT/Services.go:
```diff
@@ -726,6 +726,41 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+			// the reverse DNAT for reply packet is done at bpf_lxc. If we don't wait for endpoint regneration
+			// to complete, there may be race condition cases, where the reverse DNAT for reply packet is
+			// may not performed by both bpf_lxc and bpf_host.
+			// Hence, we need to wait for endpoint regeneration to complete on endpoints.
```
Could you move this right above the call to `WaitForCiliumPodToCompleteRegeneration`, please? That's what it's commenting.

krishgobinath replied:
The `WaitForCiliumPodToCompleteRegeneration` called twice in this test, this ends up repeating the same comment multiple times in the file.  As long as we explain why we are calling this helper function `WaitForCiliumPodToCompleteRegeneration` in a `test` wouldn't be suffice for user to understand the need for the function ?

pchaigno replied:
Adding the comment above the first occurrence should be enough. Where it is right now, folks will expect an explanation of the function, which is not what the comment is about. As an alternative, I'm also fine with adding the comment in the commit description.

krishgobinath replied:
Update comment in `test/k8sT/Services.go` as mentioned in previous comment.

joamaki commented on test/k8sT/Services.go:
```diff
@@ -1078,6 +1146,44 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
 					SkipItIf(helpers.DoesNotExistNodeWithoutCilium, "Tests externalIPs", func() {
 						testExternalIPs(kubectl, ni)
 					})
+					SkipContextIf(helpers.DoesNotRunOn54OrLaterKernel, "Test NodePort with netfilterCompatMode=true", func() {
```
It appears we have the same test code three times here. Please move it to a helper and pass in e.g. the cilium config to it. This will avoid a situation where we might fix one of these if it has issues, but forget to change the copies.

joamaki replied:
Also these tests don't actually seem to test the behavour introduced by the new flag, e.g. that netfilter actually sees the packets correctly. I would strongly recommend adding at least a rudimentary check that a netfilter rule can be exercised using this flag, and the reverse case. Otherwise nothing will stop this from breaking in the future.

krishgobinath replied:
Agreed valid point.  Added an `iptable rule`  in the test to check the packet pass through netfilter. I had used same the iptable rule to replicate the original Issue. 

Regarding, using a single helper function for moving the test [spec](https://github.com/cilium/cilium/pull/17504/files#diff-726ed9ddc3bc0d4430eaa9911b2496376f0f5a8445ff89988966eb216328ca2aR729-R767) is bit more trouble. I had tried using a helper function to move code in this spec into a helper function, somehow infra doesn't like it. At runtime, when i pass the [`kubectl`](https://github.com/cilium/cilium/blob/master/test/k8sT/Services.go#L70) var as an argument to helper function, the helper function get always `nil` value for kubectl.  Rather i prefer leave the test code as it is, current form is much more readable.

pchaigno replied:
> Regarding, using a single helper function for moving the test spec is bit more trouble. I had tried using a helper function to move code in this spec into a helper function, somehow infra doesn't like it. At runtime, when i pass the kubectl var as an argument to helper function, the helper function get always nil value for kubectl. Rather i prefer leave the test code as it is, current form is much more readable.

I disagree. It's possible to refactor this code to avoid duplicating it three times. If your refactoring causes issues, then we need to debug those issues. I'm happy to help on Slack of course!

In particular, there are two choices here I think:
1. Define a single helper function for the whole block passed to `SkipContextIf`. That function will have to take the Cilium configuration in argument and maybe a flag for the `testNodePort` statement (depends on why those differ between the different `SkipContextIf` blocks).
2. Define two helper functions, for the `BeforeAll` and `AfterAll` blocks.

krishgobinath replied:
Refactored test code, added a  new function `installCiliumAndWaitForToComplete` in `test/k8sT/service_helpers.go` to move some code in Services.go

joamaki commented on daemon/cmd/daemon_main.go:
```diff
@@ -984,6 +984,10 @@ func init() {
 	flags.Bool(option.EnableCustomCallsName, false, "Enable tail call hooks for custom eBPF programs")
 	option.BindEnv(option.EnableCustomCallsName)
 
+	flags.Bool(option.NetfilterCompatibleMode, false, "If set to true, guarantees that all traffic will pass through netfilter and that iptable rules will be enforced. This mode may reduce network throughput. If set to false (default), it does not guarantee that all traffic will pass through netfilter.")
```
Does this description actually make sense? The way I'm understanding it from the changes is that we DNAT earlier and hence netfilter would see the correct address. It doesn't seem to affect whether or not "traffic will pass through netfilter".

krishgobinath replied:
The fix in this PR, guarantees that for the nodeport traffic request/reply packets will go through netfilter. Without using `NetfilterCompatibleMode` it can't guarantee that both direction of traffic would be seen by netfilter.

pchaigno reviewed (changes requested):
Blocking on test refactoring. The rest looks good to me!

Once that is addressed, we should be good to merge (assuming CI passes) as all team review requests are covered.

kkourt reviewed (changes requested):
document the current limitations of NETFILTER_COMPAT_MODE at appropriate places, i.e., that it currently only affects nodeport traffic. See previous discussion: https://github.com/cilium/cilium/pull/17504#issuecomment-949359594

joamaki reviewed (approved):
LGTM once the test code has been cleaned up. Approving up-front as already blocked on @pchaigno's review.

qmonnet commented on bpf/bpf_host.c:
```diff
@@ -1079,6 +1079,19 @@ int to_netdev(struct __ctx_buff *ctx __maybe_unused)
+#ifdef NETFILTER_COMPAT_MODE
+
+	/* For a packet which is a reply from a local endpoint to NodePort request,
+	 * do rev-DNAT.To determine if it is Nodeport traffic, do conntrack lookup for
```
```suggestion
	 * do rev-DNAT. To determine if it is Nodeport traffic, do conntrack lookup for
```

qmonnet commented on daemon/cmd/daemon_main.go:
```diff
@@ -1008,6 +1008,15 @@ func initializeFlags() {
 
+	flags.Bool(option.NetfilterCompatibleMode, false, "If set to true, guarantees that all traffic will pass through netfilter and that iptable rules will be enforced. This mode may reduce network throughput. If set to false (default), it does not guarantee that all traffic will pass through netfilter.")
+	// Currently, NetfilterCompatibleMode guarantees only the nodeport traffic can pass through
+	// netfilter. Once its guarantees all traffic can pass through netfilter,
```
```suggestion
	// netfilter. Once it guarantees all traffic can pass through netfilter,
```

qmonnet commented on daemon/cmd/daemon_main.go:
```diff
@@ -1534,6 +1543,19 @@ func initEnv(cmd *cobra.Command) {
+
+	// Enabling NetfilterCompatibleMode increases BPF instruction count size, and may
+	// cause issues in kernels that have lower instruction complexity limit. Hence enabling,
+	// this feature only in kernels support higher instruction complexity limit (5.2 or later).
```
```suggestion
	// cause issues in kernels that have lower instruction complexity limit. Hence enabling
	// this feature only in kernels with a higher instruction complexity limit (5.2 or later).
```

qmonnet commented on daemon/cmd/daemon_main.go:
```diff
@@ -1534,6 +1543,19 @@ func initEnv(cmd *cobra.Command) {
+	if option.Config.NetfilterCompatibleMode {
+		if !probes.NewProbeManager().GetMisc().HaveLargeInsnLimit {
+			option.Config.NetfilterCompatibleMode = false
+			log.Warn("netfilter-compatible-mode is enabled only on 5.2 and higher versions of kernel.")
```
It is not _always_ enabled on recent kernels.
```suggestion
			log.Warn("netfilter-compatible-mode requires kernel version 5.2 or higher.")
```

qmonnet commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1042,6 +1042,9 @@ enableIPv6Masquerade: true
 egressGateway:
   enabled: false
 
+# -- Enables the netfilter compatible mode, guarantees that all traffic will pass through netfilter.
```
```suggestion
# -- Enable the netfilter compatible mode, guarantee that all traffic will pass through netfilter.
```

qmonnet commented on pkg/option/config.go:
```diff
@@ -985,6 +985,10 @@ const (
 	// regardless of whether it's available in the pool.
 	BypassIPAvailabilityUponRestore = "bypass-ip-availability-upon-restore"
+
+	// NetfilterCompatibleMode guarantees the traffic to pass through kernel
```
```suggestion
	// NetfilterCompatibleMode guarantees the traffic will pass through kernel
```

qmonnet commented on pkg/option/config.go:
```diff
@@ -1691,6 +1695,10 @@ type DaemonConfig struct {
 	// EnableHostLegacyRouting enables the old routing path via stack.
 	EnableHostLegacyRouting bool
 
+	// NetfilterCompatibleMode guarantees the traffic to pass through kernel
```
```suggestion
	// NetfilterCompatibleMode guarantees the traffic will pass through kernel
```

qmonnet commented on test/k8sT/service_helpers.go:
```diff
@@ -1307,3 +1307,31 @@ func testDSR(kubectl *helpers.Kubectl, ni *nodesInfo, sourcePortForCTGCtest int)
+		// WaitForCiliumPodToCompleteRegeneration guarantees that endpoint regeneration complete,
+		// and the caller may proceed further to test functionalities. For example in NetfilterCompatibleMode
+		// the reverse DNAT for reply packet is done at bpf_host. Without NetfilterCompatibleMode enabled,
+		// the reverse DNAT for reply packet is done at bpf_lxc. If we don't wait for endpoint regneration
```
```suggestion
		// the reverse DNAT for reply packet is done at bpf_lxc. If we don't wait for endpoint regeneration
```
**Note:** You also have this typo in your commit log message.

qmonnet commented on test/k8sT/service_helpers.go:
```diff
@@ -1307,3 +1307,31 @@ func testDSR(kubectl *helpers.Kubectl, ni *nodesInfo, sourcePortForCTGCtest int)
+	ciliumPods, err := kubectl.GetCiliumPods()
+	Expect(err).To(BeNil(), "Cannot get cilium pods")
+	for _, cp := range ciliumPods {
+		// WaitForCiliumPodToCompleteRegeneration guarantees that endpoint regeneration complete,
```
```suggestion
		// WaitForCiliumPodToCompleteRegeneration guarantees that endpoint regeneration completes,
```

qmonnet commented on test/k8sT/service_helpers.go:
```diff
@@ -1307,3 +1307,31 @@ func testDSR(kubectl *helpers.Kubectl, ni *nodesInfo, sourcePortForCTGCtest int)
+		// the reverse DNAT for reply packet is done at bpf_host. Without NetfilterCompatibleMode enabled,
+		// the reverse DNAT for reply packet is done at bpf_lxc. If we don't wait for endpoint regneration
+		// to complete, there may be race condition cases, where the reverse DNAT for reply packet is
+		// may not performed by both bpf_lxc and bpf_host.
```
```suggestion
		// to complete, there may be race condition cases, where the reverse DNAT for reply packet
		// may not be performed by both bpf_lxc and bpf_host.
```

qmonnet commented on test/k8sT/service_helpers.go:
```diff
@@ -1307,3 +1307,31 @@ func testDSR(kubectl *helpers.Kubectl, ni *nodesInfo, sourcePortForCTGCtest int)
+
+// Helper function to install cilium, wait for cilium pod to complete regeneration.
+// and installs a iptable rule on node Cilium-agent is running.
+// this function called in the Context of NetfilterCompatMode verification
```
```suggestion
// Helper function to install cilium, wait for cilium pod to complete regeneration,
// and install an iptable rule on the node where cilium-agent is running.
// This function is called in the context of NetfilterCompatMode verification.
```

qmonnet commented on test/k8sT/service_helpers.go:
```diff
@@ -1307,3 +1307,31 @@ func testDSR(kubectl *helpers.Kubectl, ni *nodesInfo, sourcePortForCTGCtest int)
+// Helper function to install cilium, wait for cilium pod to complete regeneration.
+// and installs a iptable rule on node Cilium-agent is running.
+// this function called in the Context of NetfilterCompatMode verification
+func installCiliumAndWaitForToComplete(kubectl *helpers.Kubectl, ciliumFileName string, options map[string]string, iptableRule string, isNetfilterCompatMode bool) {
```
`...WaitForToComplete` doesn't read well, we don't know what is supposed to complete. What about `installCiliumAndWaitForRegeneration`? (or `installCiliumAndWaitForRegenerationToComplete` but it's becoming a very long function name).

krishgobinath replied:
Updated function name to `installCiliumAndWaitForRegenerationToComplete`

qmonnet commented on test/helpers/kubectl.go:
```diff
@@ -4490,3 +4490,36 @@ func (kub *Kubectl) NslookupInPod(namespace, pod string, target string) (err err
+		cmd := fmt.Sprintf("cilium endpoint list -o jsonpath='{[?(@.status.identity.id==%d)].status.log}'", ReservedIdentityHost)
+		res := kub.CiliumExecContext(context.Background(), ciliumPod, cmd)
+		if !res.WasSuccessful() {
+			kub.Logger().Errorf("unable to run command '%s' to retrieve status log of  host endpoint from %s: %s",
```
```suggestion
			kub.Logger().Errorf("unable to run command '%s' to retrieve status log of host endpoint from %s: %s",
```

qmonnet commented on test/helpers/kubectl.go:
```diff
@@ -4490,3 +4490,36 @@ func (kub *Kubectl) NslookupInPod(namespace, pod string, target string) (err err
+
+		var logs []models.EndpointStatusChange
+		if err := res.Unmarshal(&logs); err != nil {
+			kub.Logger().Errorf("unable to unmarshal EndpointStatusChange slice '%#v': %s", logs, err)
```
Should be `%w` for errors?

```suggestion
			kub.Logger().Errorf("unable to unmarshal EndpointStatusChange slice '%#v': %w", logs, err)
```

qmonnet reviewed (changes requested):
I found a few nits, mostly typos, all should be trivial to address.

kkourt commented on Documentation/helm-values.rst:
```diff
@@ -937,6 +937,10 @@
      - string
      - ``"cilium"``
+   * - netfilterCompatMode
+     - Enable the netfilter compatible mode, guarantee that all traffic will pass through netfilter.
```
Based on our previous discussion (e.g., see: https://github.com/cilium/cilium/pull/17504#issuecomment-949359594), I believe this will end up confusing end users. I think we should make it clear in the code but also in user-facing documentation that, at least currently, this only affects nodeport traffic.

krishgobinath replied:
Updated

kkourt commented on bpf/bpf_host.c:
```diff
@@ -1079,6 +1079,19 @@ int to_netdev(struct __ctx_buff *ctx __maybe_unused)
+
+#ifdef NETFILTER_COMPAT_MODE
+
+	/* For a packet which is a reply from a local endpoint to NodePort request,
```
```suggestion
	/* For a packet which is a reply from a local endpoint to a NodePort request,
```

kkourt commented on daemon/cmd/daemon_main.go:
```diff
@@ -1008,6 +1008,15 @@ func initializeFlags() {
 	flags.Bool(option.EnableCustomCallsName, false, "Enable tail call hooks for custom eBPF programs")
 	option.BindEnv(option.EnableCustomCallsName)
 
+	flags.Bool(option.NetfilterCompatibleMode, false, "If set to true, guarantees that all traffic will pass through netfilter and that iptable rules will be enforced. This mode may reduce network throughput. If set to false (default), it does not guarantee that all traffic will pass through netfilter.")
```
```suggestion
	flags.Bool(option.NetfilterCompatibleMode, false, "If set to true, guarantees that traffic will pass through netfilter and that iptable rules will be enforced. This mode may reduce network throughput. If set to false (default), it does not guarantee that all traffic will pass through netfilter. This option only affects NodePort traffic.")
```

kkourt commented on pkg/option/config.go:
```diff
@@ -989,6 +989,10 @@ const (
 	EnableK8sTerminatingEndpoint = "enable-k8s-terminating-endpoint"
+
+	// NetfilterCompatibleMode guarantees the traffic will pass through kernel
+	// netfilter.
```
```suggestion
	// netfilter. This option only affects NodePort traffic.
```

kkourt commented on pkg/option/config.go:
```diff
@@ -1698,6 +1702,10 @@ type DaemonConfig struct {
 	EnableHostLegacyRouting bool
 
+	// NetfilterCompatibleMode guarantees the traffic will pass through kernel
+	// netfilter.
```
```suggestion
	// netfilter. Currently, it only affects NodePort traffic.
```

kkourt reviewed (approved):
LGTM! Thanks!

Weil0ng committed:
Temporarily enables nodeport tests

Signed-off-by: Weilong Cui <cuiwl@google.com>

brb commented on test/k8sT/Services.go:
```diff
@@ -568,9 +568,7 @@ var _ = SkipDescribeIf(helpers.RunsOn54Kernel, "K8sServicesTest", func() {
-	SkipContextIf(func() bool {
-		return helpers.SkipQuarantined() && helpers.RunsOnNetNextKernel()
-	}, "Checks service across nodes", func() {
+	SkipContextIf(func() bool { return helpers.RunsOnNetNextKernel() }, "Checks service across nodes", func() {
```
You need to remove all conditions for skipping.

Weil0ng replied:
confused, are these tests meant to run against net-next or not?

Weil0ng replied:
(meanwhile will remove condition as you suggested to trigger tests)

brb replied:
Yes, they are meant to be run on net-next. `SkipContextIf(func() bool { return helpers.RunsOnNetNextKernel() }` reads as "skip context if it runs on net-next kernel".

Weil0ng replied:
Yea, curious then why was that condition checked in in the first place? I mean if we are to quarantine the tests temporarily, then `SkipQuarantined` should do it? That's why I only removed that in the first place, I thought the other condition is how it's meant to run lol
